### ISSUE TITLE
COM_getLanguageFromURL does not work at topic page

### ISSUE THREAD

**Issue by ivywe** [Geeklog-Core/geeklog#819]
COM_getLanguageFromURL does not work at topic page.
COM_getLanguageFromURL retrun null,

/index.php/topic/home_en
/index.php/topic/home_ja

example:
https://demo.geeklog.jp/index.php/topic/home_en
https://demo.geeklog.jp/index.php/topic/home_ja

**Comment by eSilverStrike** [comment]
Are you referring to this issue #801 ?

**Comment by ivywe** [comment]
Yes. 
hiroron debugged COM_getLanguageFromURL() right now.
I will send the source to mystral-kk.
thanks.

**Comment by eSilverStrike** [comment]
I was just debugging it myself. You can send it to me.

**Comment by ivywe** [comment]
Thanks!
 I sent to you.

**Comment by eSilverStrike** [comment]
I really tested this fix for topics and articles with url_rewriting enabled and not. It seems to work fine. I have not tested yet with url_routing enabled though as it seems to be broken at the moment. 

I am not sure if this fix will work with url routing enabled without some tweaks especially for routes that the id of the item is not the last thing in the url.

Once it has been tested with url_routing enabled and everything works we can close this issue.

**Comment by eSilverStrike** [comment]
@hirorongl COM_getLanguageFromURL, COM_getLanguageIdFromURL does not work with URL Routing enabled.

With Geeklog articles, topics, staticpages (blocks do as well but you cannot access a block via a url) are the only thing that uses _en in the id itemid_en to determine the language (not sure if any 3rd party plugin supports this but COM_getLanguageFromURL doesn't care about this

For "rewritten" URLs we assume that the first parameter after the script name is the ID, e.g. 
/article.php/story-id-here_en 
/index.php/topic/home_ja from /index.php?topic=home_ja

For routing and depending on the setting (with or without index.php), this

http://yourdomain/article.php?story=@sid

gets turned into by default

/article/@sid or index.php/article/@sid

and if the webmaster edits the routing you could even get routes similar to the format (not sure why they would of this but it is possible)

/a/@sid or index.php/a/@sid
/a/b/@sid or index.php/a/b/@sid
/@sid/article or index.php/@sid/article/ (notice id is first in  this example)
/a/@sid/b/c or index.php/a/@sid/b/c

@mystralkk @hirorongl as far as I know COM_getLanguageFromURL only gets used in switchlang.php but since it is a COM_ function I assume some 3rd party plugins may call it.

COM_getLanguageFromURL should support URL routing. If it does then the only way this will work is if url routing is enabled, COM_getLanguageFromURL  needs to access the url routing records to see if there are any route matches so it can pull the item id from the correct variable.

Does anyone have any thoughts on this or other ideas?

**Comment by eSilverStrike** [comment]
We should also maybe lock down he COM_getLanguageIdFromURL function (and/or maybe switchlang.php) to return a value only if the url supports it because technically other plugin ids could have _en, _ja, etc...  in the id and not even be referring to multiple language support.

This means this function should work only for articles, topics, and staticpage URLs. We should also add a new plugin api that indicates support for this function (or maybe multi-language in general). The staticpage plugin would be updated to use this new api.

**Comment by eSilverStrike** [comment]
Looking at COM_getLanguageIdFromURL it has many issues. The 4 main glaring ones are

1) No way to determine what the url is for (ie topic, article, staticpage, other plugin that supports this feature)
2) No way to 100% determine what variable is the id or if it even exists (for regular url, rewritten url, route url)
3) No way to 100% determine if the url even supports multiple languages (for regular url, rewritten url, route url)


Currently COM_getLanguageIdFromURL handles this by guessing and searching (either a regular url or rewrite url) for an underscore following by 2 characters at the end of the first url variable. With url rewrite enabled it then attempts to refine this guess. This of course can run into problems since Geeklog in no way locks down the first variable from containing an underscore in the 3rd last character for Core or any plugin. This means false positives are very possible. So no matter how much we try to lock it down

In a perfect world we should have the plugin register with the URL class if it supports this feature and if it does allow the URL class to be able to retrieve this information (being the id of the item being displayed and  the language id it is for) easily for any type of URL (normal, rewrite, routing). If we do it this way we may break 3rd party plugin which support this feature (I don't know of any that do)

The other problem with COM_getLanguageIdFromURL is that a URL variable can be passed into the function which seems to indicate that the URL other than the current one can be used (which we never do as far as I can tell). I think we should remove this passed url variable from the function and also rename the function. Functions that start with COM_ indicate that it can be called by other plugins if needed. This function should not be called by other plugins. Plugins should only use COM_getLanguage if they need the language.

This is also true with COM_getLanguageFromURL I think. Plugins should not be calling this either and it should just use the current URL and not allow one to be passed in.

So for functions that exist in lib-common.php that should NOT be called by plugins, only by core itself, what should they start with to indicate this status? CORE_getLanguageIdFromURL or maybe PRIV_getLanguageIdFromURL or nothing _getLanguageIdFromURL?

Hopfully this all makes sense and I haven't messed up any logic. Anyone have any other thoughts on this?

Either we do what I suggest or we leave it as is with a few minor changes to COM_getLanguageIdFromURL to hopefully make it better at guessing the language.

**Comment by eSilverStrike** [comment]
- COM_getLanguageFromURL renamed to _getLanguageFromURL. Now should only be called by core
- Unfortunately COM_getLanguage needs to get called  almost right away when a visitor visits a page. This means if the language is not already set from a previous visit _getLanguageFromURL is required to guess at the language. Code still needs to be added to guess the language id for url_routing
- Fix for switch language block to not display if user not allowed to change language.
- Added to $_URL class that allows item (like a article or plugin) to set the id for the url and the language. This means no guessing by switch language block anymore
- New code for plugins that uses $_URL class to support an item that has multiple languages (like topics, articles, staticpage). This may break multi-language support a bit for any other plugins that support this type of feature 
// If user is allowed to switch languages
```
if ($_CONF['allow_user_language'] == 1) {
    // Let's figure out if page is for specific language
    // If so let URL class know in case user changes language
    $page_lang = COM_getLanguageIdForObject($page);
    if (!empty($page) AND !empty($page_lang)) {
        $_URL->setItemInfo($page, $page_lang);
    }
}
```

**Comment by eSilverStrike** [comment]
So any function in lib-common.php that starts with a underscore _ SHOULD only be called by Geeklog Core.

So the only thing that needs fixing now is url routing.

@mystralkk The _getLanguageFromURL function still needs some code to figure out url_routing. Is there any way that the routing class could return this information? (it is the best spot for it). The dispatch function figures out what the original url_routing url is it looks like. When you have time :-) could you update the function (or create a new one that could possible return the id of the url if it exists (ie it is not a normal redirect or something) along with what the url route is for (like article, topic, staticpage, or something else)

**Comment by eSilverStrike** [comment]
Crap... working on a slightly different solution that may work out better. @mystralkk hold off on doing anything

**Comment by eSilverStrike** [comment]
@mystralkk Okay I think I have solved this. The main issue  is COM_getLanguage is called really early when the page loads and at this point the language id of the url is figured out. This means that plugins functions.inc is not loaded which means I have no access to plugin functions or their configuration which makes it difficult for the plugin to tell me what to look for if it supports multiple languages.

I have found a way around this.  What I can do is on install have the plugin insert a config option for Core. Core Config options are loaded right away (along with what plugins are enabled). This core config option would be hidden and contain the plugin name in the config option name like so:

```
    $c->add('langurl_article',array('', 'article.php', 'story'),'@hidden',1,7,NULL,1400,TRUE, $me, 7);
    $c->add('langurl_topic',array('', 'index.php', 'topic'),'@hidden',7,31,1,1830,TRUE, $me, 31);
    $c->add('langurl_staticpages',array('staticpages', 'index.php', 'page'),'@hidden',7,31,1,1830,TRUE, $me, 31);
```

So basically I just have to loop through these config options to find the one that matches the current url. If it does Geeklog knows then to check for _en language marker.

If the plugin is uninstalled it would then just have to add an extra uninstall step to remove its core langurl_ config option. I also think I can get this working without any other changes to the URL and router class. 

What do you think? I don't think it's a huge issue that a plugin will be adding a hidden core config option.

**Comment by mystralkk** [comment]
Your solution sounds very nice to me.

**Comment by eSilverStrike** [comment]
Well after several rewrites I think I got it.

Multi-Language Items by plugins, articles and topics fully supported including an anonymous user directly visiting a muti language page and any user using the switch language block. Switch Language block also uses switchlang_homepage config option if enabled and item doesn't exist in language will return to homepage else a 404 is displayed. This supports normal, rewrite , and routing URLs.

No more guessing at URLs for item ids and language variables. Plugins now need to register a core hidden config option  langurl_pluginname to support multi-language items. This config option will store an array with the follow information about the url that supports multiple languages:

plugin name, directory, filename, id variable name

Also required to support PLG_getItemInfo (which calls plugin_getiteminfo_pluginname) that can return if the Id exists or not of the item.

Plugin Developers see function _getLanguageInfoFromURL in lib-common for more information (which replaced COM_getLanguageFromURL).

Topics, Articles and Staticpages Plugin fully support this feature.
